pyro/distributions/grouped_normal_normal.py

```diff
@@ -0,0 +1,142 @@
+    Normal likelihood is controlled by `obs_scale`, which is allowed to vary from observation to
+    observation. The tensor of indices `group_idx` connects each observation to one of the groups
+    specified by `prior_loc` and `prior_scale`.
+
+    See e.g. Eqn. (55) in ref. [1] for relevant expressions in a simpler case with scalar `obs_scale`.
+
+    Example:
+
+    >>> num_groups = 3
+    >>> num_data = 4
+    >>> prior_loc = torch.randn(num_groups)
+    >>> prior_scale = torch.rand(num_groups)
+    >>> obs_scale = torch.rand(num_data)
+    >>> group_idx = torch.tensor([1, 0, 2, 1]).long()
+    >>> values = torch.randn(num_data)
+    >>> gnn = GroupedNormalNormal(prior_loc, prior_scale, obs_scale, group_idx)
+    >>> assert gnn.log_prob(values).shape == ()
+
+    References:
+    [1] "Conjugate Bayesian analysis of the Gaussian distribution," Kevin P. Murphy.
+
+    :param torch.Tensor prior_loc: Tensor of shape `(num_groups,)` specifying the prior mean of the latent
+        of each group.
+    :param torch.Tensor prior_scale: Tensor of shape `(num_groups,)` specifying the prior scale of the latent
+        of each group.
+    :param torch.Tensor obs_scale: Tensor of shape `(num_data,)` specifying the scale of the observation noise
+        of each observation.
+    :param torch.LongTensor group_idx: Tensor of indices of shape `(num_data,)` linking each observation to one
+        of the `num_groups` groups that are specified in `prior_loc` and `prior_scale`.
+    """
+    arg_constraints = {
+        "prior_loc": constraints.real,
+        "prior_scale": constraints.positive,
+        "obs_scale": constraints.positive,
+    }
+    support = constraints.real
+
+    def __init__(
+        self, prior_loc, prior_scale, obs_scale, group_idx, validate_args=None
+    ):
+        prior_loc, prior_scale = broadcast_all(prior_loc, prior_scale)
+        obs_scale, group_idx = broadcast_all(obs_scale, group_idx)
+
+        self.prior_loc = prior_loc
+        self.prior_scale = prior_scale
+        self.obs_scale = obs_scale
+        self.group_idx = group_idx
+        batch_shape = prior_loc.shape[:-1]
+
+        if batch_shape != torch.Size([]):
+            raise ValueError("GroupedNormalNormal only supports trivial batch_shape's.")
+
+        self.num_groups = prior_loc.size(0)
+        if group_idx.min().item() < 0 or group_idx.max().item() >= self.num_groups:
+            raise ValueError(
+                "Each index in group_idx must be an integer in the inclusive range [0, prior_loc.size(0) - 1]."
+            )
+
+        self.num_data_per_batch = prior_loc.new_zeros(self.num_groups).scatter_add(
+            0, self.group_idx, prior_loc.new_ones(self.group_idx.shape)
+        )
+        super().__init__(batch_shape, validate_args=validate_args)
+
+    def expand(self, batch_shape, _instance=None):
+        raise NotImplementedError
+
+    def sample(self, sample_shape=()):
+        raise NotImplementedError
```

Comment:
Gosh it seems a shame not to implement at least a generator for test data
```python
has_rsample = True
def rsample(self, sample_shape=torch.Size()):
    group_part = Normal(self.prior_loc, self.prior_scale).rsample(sample_shape)
    obs_part = Normal(0, self.obs_scale).rsample(sample_shape)
    return group_part[..., self.group_idx] + obs_part
```
but I guess it's fine to omit it until someone needs it 🤔

Reply:
yeah the thing is the only distribution that a user is actually likely to care about is the posterior predictive and not the prior predictive